src/main/java/com/rsr5/gocd/riemann_notifier/GoNotificationPlugin.java

```diff
@@ -138,23 +144,68 @@ private GoPluginApiResponse handleNotificationsInterestedIn() {
         return renderJSON(SUCCESS_RESPONSE_CODE, response);
     }
 
-    private String service(JsonObject json) {
-        JsonObject pipelineObject, stageObject;
-        pipelineObject = (JsonObject) json.get("pipeline");
-        stageObject = (JsonObject) pipelineObject.get("stage");
+    private String safeGetAsString(JsonObject json, String property) {
+        JsonElement element = json.get(property);
+        if (element != null && element.isJsonPrimitive()) {
+            return element.getAsString();
+        }
+        return null;
+    }
+
+    final class StageData {
+        public String pipeline;
+        public String stage;
+        public String state;
+        public String counter;
+        public Instant start;
+        public Instant end;
+        public List<JobData> jobs;
```

Comment:
Looks like this class is in the spirit of an "immutable struct" - how about making these fields all `final`?